pkg/apis/istio/v1alpha1/istioinstaller_types.proto

```diff
@@ -0,0 +1,441 @@
+
+    // Global k8s defaults. Can be overridden per component.
+    KubernetesResourcesSpec k8s_defaults = 80;
+
+    // URL for the install package. e.g.
+    //   http://github.com/istio/istio/releases/install/1.1.2
+    //   http://github.com/istio/istio/releases/install/lts
+    //   file://tmp/istio-installer/nightly
+    // Uses compiled in install package by default.
+    string custom_package_path = 100;
+    // Root for docker image paths e.g. docker.io/istio-release.
+    // Releases are published to docker hub under 'istio' project.
+    // Daily builds from prow are on gcr.io, and nightly builds from circle on docker.io/istionightly
+    string hub = 101;
+    // Version tag for docker images e.g. 1.0.6
+    string tag = 102;
+}
+
+// Configuration options for traffic management.
+message TrafficManagementFeatureSpec {
+    // Selects whether traffic management is installed.
+    // Must be enabled to enable any sub-component.
+    google.protobuf.BoolValue enabled = 1;
+    // Namespace that all traffic management components are installed into.
+    string namespace = 2;
+
+    // Component specific config.
+    PilotComponentSpec pilot = 30;
+    ProxyComponentSpec proxy = 31;
+    SidecarInjectorComponentSpec sidecar_injector = 32;
+}
+
+// Configuration options for the policy and telemetry feature.
+message PolicyTelemetryFeatureSpec {
+    // Default behavior of the sidecar for handling outbound traffic from the application.
+    enum OutboundPolicy {
+        // Outbound traffic to unknown destinations will be allowed, in case there are no
+        // services or ServiceEntries for the destination port
+        ALLOW_ANY = 0;
+        // Restrict outbound traffic to services defined in the service registry as well
+        // as those defined through ServiceEntries
+        REGISTRY_ONLY = 1;
+    }
+
+    // If set, allows traffic in cases when the mixer policy service cannot be reached.
+    google.protobuf.BoolValue policy_check_fail_open = 10;
+    // Selects the outbound traffic policy mode.
+    OutboundPolicy outbound_traffic_policy_mode = 11;
+
+    PolicyComponentSpec policy = 50;
+    TelemetryComponentSpec telemetry = 51;
+}
+
+// Configuration options for security feature.
+message SecurityFeatureSpec {
+    // Selects whether security feature is installed. Must be set for any sub-component to be installed.
+    google.protobuf.BoolValue enabled = 1;
+
+    // Namespace that security components are installed into.
+    string namespace = 2;
+
+    // Enables MTLS for control plane.
+    // Will result in delays starting the pods while secrets are propagated, not recommended for tests.
+    google.protobuf.BoolValue control_plane_mtls = 13;
+
+    // Enables MTLS for service to service traffic.
+    google.protobuf.BoolValue data_plane_mtls = 14;
+
+    CitadelComponentSpec citadel = 20;
+    CertManagerComponentSpec cert_manager = 21;
+    NodeAgentComponentSpec node_agent = 22;
```

Comment:
having these components spec here and the need for user to understand what each component is before they can configure things is hard to use imho and not much difference than what we offer from helm values today.

Reply:
The issue is that k8s resources and overlays are necessarily a component rather than feature concept. keep in mind that these are overrides over defaults - most users won't need to worry about this level of detail. 
The initial API did in fact migrate out all the values.yaml configuration into a structured API, but the review feedback  was that the Istio config part was going into component CRDs and therefore shouldn't be a part of this installation API. 
Se the idea here is to keep values.yaml but provide an alternative API just for that parts that will not go into the future CRDs i.e. k8s level stuff.
Having said that, we may want to revisit this path and consider only using values.yaml only under the hood rather than exposing directly, and go back to the initial proposal for a complete API replacement of values.yaml.

Reply:
I think it would be better if we could separate the functional details from resourcing requirements.  While that might be a bit messier from a configuration structure, I think separation is beneficial.  For the work we've been doing with maistra, we've started to use a simple config for features, with a small defaultResources setting, which allows for a better ootb experience on developer machines (things like grafana can be resource hogs).

Reply:
Resourcing must be tied to components so I'm interpreting the comment as suggesting splitting the components from the features. The reason we probably don't want to do that is that components are logically grouped so they can easily share a namespace (and enablement) and all the policy control associated with the feature/namespace. We want to be able to control all that at feature level, with the ability to override per component.

Reply:
@rcernich do you have some example yamls for separating the functional details from resourcing requirements?  I think that will help unclutter things and not exposing our impl details to users who only care about function or feature configuration.

Reply:
I actually don't and based on @ostromart's last comment, maybe what I was thinking is less useful.  I do think it's beneficial to segregate resourcing configuration from feature configuration, but maybe that can be done by simply having something like "ResourceConfig" as part of each section.  The only details within that section would be things like resource limits, node selectors, autoscaling details, etc.

A generic mock up:
```yaml
componentA:
  featureA:
    ...
  operationalConfig:
    resourceRequirements:  # k8s ResourceRequirements
    podDisruptionBudget:  # k8s PodDisruptionBudget
    autoScaling: # things like min/max replicas, target cpu utilization, etc.
    nodeSelector: # map of label->value

operationalConfig: # same as above, but global defaults, applied when no component.operationalConfig is specified.
```

The above is a crude example and assuming one deployment resource per component, but hopefully, it conveys the general idea.

Reply:
Thanks @rcernich - I think this starts with component first and list features inside each component, which is component centric.

I 'd prefer to be feature centric so user doesn't have to understand component if he/she doesn't need to.  I had thought you were proposing seperating the component spec out of the feature spec, e.g. something like this:

```
  traffic_management:
    enabled: true
    ...
  components:
    pilot:
        k8s:
          resources:
            limits:
              cpu: 444m
              memory: 333Mi
            requests:
              cpu: 222m
              memory: 111Mi
          readinessProbe:
            failureThreshold: 44
            initialDelaySeconds: 11
            periodSeconds: 22
            successThreshold: 33
```

Reply:
Keep in mind that all of this API is optional - users would not normally dive down to component level, only if they had to (say) modify k8s resources. Defaults are always used if something is not defined and that would be the case most of the time. 
As far as splitting out the components outside of the feature groups, the disadvantage of doing that is that some settings (e.g. install namespace, enabled) are inherited from the feature level and it makes sense imo to maintain that relationship.

Reply:
@linsun, your example is close to what I was suggesting, except that I would move `components` under `traffic_management`, and possibly call it `operationalConfig`.  @ostromart, regarding namespaces, IMO, those are operational concerns, so I would place those under `components`/`operationalConfig` too.  That said, I don't know what the technical concern is regarding splitting up the control plane across multiple namespaces (i.e. why you'd want to do that).

Reply:
This may be a dumb question, but if different users are administering different aspects of the control plane and in different namespaces, why would we configure all of them with a single CR?  If they are truly operationally owned by different groups, then they shouldn't be part of the same CR, as any operational owner will have the ability to modify the CR, including those aspects for which they are not responsible.

Furthermore, I think you need to consider the security implications of allowing somebody to specify a namespace that they may not have access to.  (We're working on this problem in Maistra, regarding how we ensure random namespaces can't be joined into a mesh in a multitenant environment.)

It may be important to point out that one of the purposes of using an operator is that the operator is the tool that has the privileges to create all the necessary resources, giving regular users the ability to install a control plane (i.e. joe user doesn't need cluster-admin, he just needs permissions to create an istio control plane cr).

Reply:
That's a fair question - this is actually another reason for separating config concerns from install concerns. If you lump all that together into the operator CRD and put it in its own namespace, feature/component administrators lose control over things they _should_ be allowed to configure. That's a good argument for moving the config out of the operator. You can still argue that some things like pilot resources should belong to the traffic management domain but in practice that administrative boundaries is probably ok because install parameters tend to be fairly static, so operators would likely be ok with going through a process to change them rather than needing direct access.

Reply:
While I asked the question, I actually believe that the CR be the place where the configuration is exposed to the user.  If the user modifies the configuration, the operator can make sure the changes are rolled out appropriately.

Reply:
That's not the WG consensus (cc @costinm, @linsun). I personally spent a lot of time and energy creating a very complete config schema for the operator, but that's not the direction Istio is going, and I understand the rationale. So I'm ok with the decision to keep just installation footprint definition in the operator and move everything else to what is currently called mesh config. Again, I think this is a broader topic and this is not the right place for that discussion - feel free to put it on the agenda for the WG weekly or raise it at the TOC. 
Closing here in the understanding that we can continue in the other forums.